Ported simulator to ROS2

lordreas opened the pull request:
This PR brings the simulator to ROS2. I have 

- renamed paths from gazebo_ignition to gz
- replaced gazebo ignition with GZ Harmonic
- added a submodule for a mavlink plugin for gz, as it is not available out of the box: `srl_sim_gz/px4-gzsim-plugins`
- ported launch files, CMakeLists, and package.xml
- updated the PX4-Autopilot to a newer version
- ported the sim resources to be compatible with the new autopilot and GZ
- updated the okvis configs 

I tested everything with the srl_navigation stack and standalone, just launching it. I have not yet tested srl_mpc_examples as I havent set up mav_ekf

lordreas committed:
renamed srl_sim_gazebo_ignition to srl_sim_gz

lordreas committed:
ported launch files to ros 2

lordreas committed:
ported package.xml and CMakesList to ros2

lordreas committed:
ported sim ressources to gz

lordreas committed:
ported srl_sim_util to ros2

lordreas committed:
updated bash launch scripts for gz and px4

lordreas committed:
ported rmf_owl launch file to new okvis version

lordreas committed:
updated .gitmodules

lordreas committed:
fixed simulation okvis config for rmf_owl

lordreas committed:
fully ported px4 arming code for offboard operation to ros2

Commit:
fixed missing submodule and config file

Commit:
ported srl_mpc_examples.

Tested up to trajectory publisher, but not other two launch files, because I don't have mav_ekf atm.

Commit:
updated readmes

SebsBarbas commented on srl_sim_gz/launch/gazebo.launch.xml:
```diff
@@ -9,24 +9,23 @@
-    <arg name="robot_name" value="$(arg airframe)"/>
+  <include file="$(find-pkg-share srl_sim_gz)/launch/rmf_owl_ros_ign_bridges.launch.xml">
+    <arg name="world"      value="$(var world)"/>
+    <arg name="robot_name" value="$(var airframe)"/>
```
Could we add in this launch file setting the environment variable so that it does not communicate with the exterior? I can imagine that this knowledge will be lost/someone will launch this and forget to set the environment variable and it will screw up the other people using ROS2.

lordreas replied:
When I tested it quickly yesterday, it didn't work. I guess this env needs to be set before any ros is launched... But I didn't have time to investigate further. Will do this after my vacation (end of july).

jaehyungjung commented on srl_sim_gz/launch/digiforest_sim.launch.xml:
```diff
@@ -1,7 +1,7 @@
 <launch>
   <!-- The world to load in Gazebo. It must contain the MAV model. -->
-  <arg name="world" default="$(find srl_sim_gazebo_ignition)/resources/worlds/depot.world"/>
+  <arg name="world" default="$(find srl_sim_gz)/resources/worlds/depot.world"/>
```
This one seems depot world, not the forest environment. Is it okay?

lordreas replied:
Yes, I kept it as it was before. Depot is kind of the default simulator we are using for planning and mpc developments. That's why I would keep it the default.

jaehyungjung commented on srl_sim_gz/config/rmf_owl_okvis.yaml:
```diff
@@ -12,7 +12,8 @@ cameras:
         cam_model: pinhole,
         camera_type: gray+depth,
-        mapping: true}
+        mapping: true,
```
Did you also test with depth submap alignment and rviz visualizes the aligned points messages?

lordreas replied:
No I haven't. I'm also not really familiar with it. I can test it after my vacation (end of july) or if you have a free moment, feel free to test it with the task/lf/port2ro2 branch of srl_navigation.

chutsu commented:
Any chance this PR update can include a docker container? so that #1 can be closed :P

It would also help with containing / debugging the dev environment

sotpapathe commented on srl_sim_gz/src/gazebo.bash:
```diff
@@ -12,8 +12,8 @@ then
-ign gazebo --force-version 6 ${HEADLESS+-s} -r "$2"
+export GZ_SIM_RESOURCE_PATH=$px4_dir/../../../srl_sim_gz/share/srl_sim_gz/resources/:$GZ_SIM_RESOURCE_PATH
+export GZ_SIM_SYSTEM_PLUGIN_PATH=$px4_dir/../../../srl_sim_gz/share/srl_sim_gz/:$GZ_SIM_SYSTEM_PLUGIN_PATH
+gz sim ${HEADLESS+-s} -r $world
```
`$world` -> `"$world"` to allow handling values with spaces in them.

lordreas replied:
thanks for pointing it out, I corrected it!

sotpapathe commented on srl_sim_util/src/px4_arm_offboard.py:
```diff
@@ -2,30 +2,44 @@
+        arm_future = arm_client.call_async(arm_req)
+        rclpy.spin_until_future_complete(node, arm_future)
+        if arm_future.result() is not None and arm_future.result().success:
+            break
```
I find having the condition in the `while` much more readable that `while True:` and breaking internally.

lordreas replied:
resolved in #4

sotpapathe commented on srl_sim_util/src/px4_arm_offboard.py:
```diff
@@ -2,30 +2,44 @@
+        mode_future = mode_client.call_async(mode_req)
+        rclpy.spin_until_future_complete(node, mode_future)
+        if mode_future.result() is not None and mode_future.result().mode_sent:
+            break
```
I find having the condition in the `while` much more readable that `while True:` and breaking internally.

lordreas replied:
resolved by #4

sotpapathe commented on srl_sim_gz/resources/robots/rmf_owl/model.sdf:
```diff
@@ -34,8 +34,7 @@
           <always_on>true</always_on>
-          <topic>imu</topic>
-          <update_rate>200</update_rate>
+          <update_rate>250</update_rate>
```
Is there a reason why the IMU rate was changed? Just pointing it out in case it was by accident.

lordreas replied:
sorry this was a leftover from debugging. Thanks for pointing it out!

sotpapathe reviewed (changes requested):
I don't really have experience with ROS2 and I couldn't run it on my machine (no ROS2). I left a few comments on things that stood out to me. Feel free to ignore all of them except the one about adding double quotes around `$world`.

BoMicha committed:
fix ros2 callback logic of px4_arm_offboard.py node

In ROS 2, the executor must run in a separate thread to allow the main thread to block with rate.sleep().
Without spinning in the background, callbacks (e.g., service responses) won't be processed.
See: https://docs.ros.org/en/jazzy/How-To-Guides/Migrating-from-ROS1/Migrating-Python-Package-Example.html

bochsim committed:
Linking correct commit in mav_interface_msgs to work with srl_navigation (missing conversions)

lordreas committed:
Merge pull request #4 from ethz-mrl/px4_arm_port2ros2_fix

fix ros2 callback logic of px4_arm_offboard.py node

Commit:
corrected baud rate in model.sdf of rmf_owl

Commit:
improved gazebo.bash according to review

Commit:
updated readme for more complete install instructions

Commit:
[BUGFIX] removed blocky exposure artefacts

There is a bug in the current gz sim, where activating the <extrinsics> tag for a sensor, blocky exposure artefacts are introduced. This fix removes the tag and adjusts the okvis config to represent the perfect extrinsics resulting from having perfect center points and focal lengths.

Feel free to undo this commit once the bug in gz is fixed

Commit:
minor updates to use correct plugins following new gz nomenclature
